core/src/autogluon/core/utils/feature_selection.py

```diff
@@ -0,0 +1,537 @@
+        expected_single_feature_time = self.compute_expected_fi_time_single(X, self.model_predict_time, n_subsample, n_sample)
+        auto_threshold = len(prioritized) > 0
+        features = sort_features_by_priority(features, prioritized, prev_importance_df, using_prev_fit_fi)
+
+        # if we do not have enough time to evaluate feature importance for all features, do so only for some (first n_evaluated_features elements of features)
+        n_evaluated_features = max([i for i in range(0, n_features+1) if i * expected_single_feature_time <= time_budget])
+        if n_evaluated_features == 0:
+            prune_time = time.time() - time_start
+            self.time_limit = self.time_limit - prune_time
+            self._fi_time_elapsed = self._fi_time_elapsed + prune_time
+            return features, unevaluated_fi_df_template(features), prune_time
+        evaluated_features = features[:n_evaluated_features]
+        unevaluated_features = features[n_evaluated_features:]
+        time_budget_fi = time_budget - (time.time() - time_start)
+        logger.log(15, f"\tComputing feature importance for {n_evaluated_features}/{n_features} features with {n_shuffle} shuffles.")
+        fi_kwargs = {'X': X, 'y': y, 'num_shuffle_sets': n_shuffle, 'subsample_size': n_subsample, 'features': evaluated_features,
+                     'time_limit': time_budget_fi, 'silent': True}
+        fi_kwargs.update({'is_oof': True} if self.is_bagged else {})
+        evaluated_df = model.compute_feature_importance(**fi_kwargs)
+        if self.is_bagged:
+            evaluated_df['n'] = evaluated_df['n'] // len(model.models)
+
+        # if we could not compute feature importance for all features and previous feature importance estimates exist, use them
+        importance_df = pd.concat([evaluated_df, unevaluated_fi_df_template(unevaluated_features)])
+        importance_df = merge_importance_dfs(prev_importance_df, importance_df, using_prev_fit_fi)
+
+        # if auto_threshold, threshold is the mean of noise column importance score
+        if auto_threshold:
+            noise_rows = importance_df[importance_df.index.isin(prioritized)]
+            importance_df = importance_df.drop(prioritized)
+            prune_threshold = noise_rows['importance'].mean()
+
+        # use importance_df to generate next candidate features
+        time_budget_select = time_budget - (time.time() - time_start)
+        candidate_features = self.compute_next_candidate_given_fi(importance_df, prune_threshold, prune_ratio, time_budget_select, weighted)
+
+        # if noise columns exist, they should never be removed
+        if auto_threshold:
+            candidate_features = candidate_features + list(prioritized)
+            importance_df = pd.concat([importance_df, noise_rows])
+
+        feature_selection_time = time.time() - time_start
+        self.time_limit = self.time_limit - feature_selection_time
```

Comment:
This is dangerous. You may want to instead keep track of self.time_limit as a self.time_start, then you can compare time_elapsed as `time_elapsed = time.time() - self.time_start`, `time_left = self.time_limit - time_elapsed`, that way you never edit time_limit.